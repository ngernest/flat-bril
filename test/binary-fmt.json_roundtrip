{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "funcs": [
            "printBinary"
          ],
          "labels": [],
          "op": "call"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "n",
            "zero"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "funcs": [],
          "labels": [
            "end",
            "rec"
          ],
          "op": "br"
        },
        {
          "label": "rec"
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "v0",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v1"
          ],
          "funcs": [
            "printBinary"
          ],
          "labels": [],
          "op": "call"
        },
        {
          "args": [
            "v0"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "label": "end"
        }
      ],
      "name": "printBinary"
    },
    {
      "args": [
        {
          "name": "a0",
          "type": "int"
        },
        {
          "name": "a1",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a0",
            "a1"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "a1"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a0",
            "v1"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v2"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
