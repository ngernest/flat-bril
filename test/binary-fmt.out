@main
	op: call 		args: ["n"]		funcs: ["printBinary"]
@printBinary
	op: const		dest: zero		type: int		value: 0	
	op: eq   		dest: cond		type: bool		args: ["n", "zero"]	
	op: br   		args: ["cond"]		labels: ["end", "rec"]
.rec
	op: const		dest: two		type: int		value: 2	
	op: call 		dest: v0		type: int		args: ["n", "two"]		funcs: ["mod"]
	op: div  		dest: v1		type: int		args: ["n", "two"]	
	op: call 		args: ["v1"]		funcs: ["printBinary"]
	op: print		args: ["v0"]	
.end
@mod
	op: div  		dest: v0		type: int		args: ["a0", "a1"]	
	op: mul  		dest: v1		type: int		args: ["v0", "a1"]	
	op: sub  		dest: v2		type: int		args: ["a0", "v1"]	
	op: ret  		args: ["v2"]	
